File: study/lua/src/app/scene/view/countryboss/FightBtn.lua

--冷却时间帮助类
local FightBtn = class("FightBtn")

function FightBtn:ctor(textImage, timeLabel, getEndTimeFunc, timeLabelBg)
    self._textImage = textImage
    self._timeLabel = timeLabel
	self._getEndTimeFunc = getEndTimeFunc
	self._timeLabelBg = timeLabelBg
end

function FightBtn:update()
	self._canAttackTime = self._getEndTimeFunc()
	local curTime = G_ServerTime:getTime()
    -- logError("time1  "..G_ServerTime:getTimeString(self._canAttackTime))
    -- logError("time2  "..G_ServerTime:getTimeString(self.curTime))
	if curTime <= self._canAttackTime then
		self:_startCd()
		self._timeLabel:setString(G_ServerTime:getLeftSecondsString(self._canAttackTime, "00:00:00"))
		local UIActionHelper = require("app.utils.UIActionHelper")
		local action = UIActionHelper.createUpdateAction(function()
			self:_cdUpdate()
		end, 0.5)
		self._timeLabel:runAction(action)
	else
		self:_stopCd()
	end
end

function FightBtn:_stopCd()
	self._timeLabel:stopAllActions()
	self._timeLabel:setVisible(false)
	-- self._textImage:setPositionY(78)

	if self._timeLabelBg then
		self._timeLabelBg:setVisible(false)
	end
end

function FightBtn:_startCd()
	self._timeLabel:stopAllActions()
	self._timeLabel:setVisible(true)
	-- self._textImage:setPositionY(92)

	if self._timeLabelBg then
		self._timeLabelBg:setVisible(true)
	end
end

function FightBtn:_cdUpdate()
	local curTime = G_ServerTime:getTime()
	if curTime > self._canAttackTime then
		self:_stopCd()
	else
		self._timeLabel:setString(G_ServerTime:getLeftSecondsString(self._canAttackTime, "00:00:00"))
	end
end

function FightBtn:canFight()
    local curTime = G_ServerTime:getTime()
    local canAttackTime = self._getEndTimeFunc()
    if curTime > canAttackTime then
        return true
    end
    return false
end

return FightBtn

File: study/ts/script/scene/view/countryboss/FightBtn.ts
import { G_ServerTime } from "../../../init";
import UIActionHelper from "../../../utils/UIActionHelper";

const { ccclass, property } = cc._decorator;
@ccclass
export default class FightBtn extends cc.Component {
    _textImage: cc.Sprite;
    _timeLabel: cc.Label;
    _getEndTimeFunc: any;
    _timeLabelBg: cc.Sprite;
    _canAttackTime: any;

    public static newIns(textImage, timeLabel, getEndTimeFunc, timeLabelBg) {
        var comp = new cc.Node().addComponent(FightBtn);
        comp.ctor(textImage, timeLabel, getEndTimeFunc, timeLabelBg);
        return comp;
    }

    ctor(textImage, timeLabel, getEndTimeFunc, timeLabelBg) {
        this._textImage = textImage;
        this._timeLabel = timeLabel;
        this._getEndTimeFunc = getEndTimeFunc;
        this._timeLabelBg = timeLabelBg;
    }
    update() {
        this._canAttackTime = this._getEndTimeFunc();
        var curTime = G_ServerTime.getTime();
        if (curTime <= this._canAttackTime) {
            this._startCd();
            this._timeLabel.string = (G_ServerTime.getLeftSecondsString(this._canAttackTime, '00:00:00'));
            var action = UIActionHelper.createUpdateAction(function () {
                this._cdUpdate();
            }.bind(this), 0.5);
            this._timeLabel.node.runAction(action);
        } else {
            this._stopCd();
        }
    }
    _stopCd() {
        this._timeLabel.node.stopAllActions();
        this._timeLabel.node.active = (false);
        this._textImage.node.y = (27);
        if (this._timeLabelBg) {
            this._timeLabelBg.node.active = (false);
        }
    }
    _startCd() {
        this._timeLabel.node.stopAllActions();
        this._timeLabel.node.active = (true);
        this._textImage.node.y = (14 + 27);
        if (this._timeLabelBg) {
            this._timeLabelBg.node.active = (true);
        }
    }
    _cdUpdate() {
        var curTime = G_ServerTime.getTime();
        if (curTime > this._canAttackTime) {
            this._stopCd();
        } else {
            this._timeLabel.string = (G_ServerTime.getLeftSecondsString(this._canAttackTime, '00:00:00'));
        }
    }
    canFight() {
        var curTime = G_ServerTime.getTime();
        var canAttackTime = this._getEndTimeFunc();
        if (curTime > canAttackTime) {
            return true;
        }
        return false;
    }
}